File: 407-trapping-rain-water-ii/trapping-rain-water-ii.java
class Solution {
    class Block implements Comparable<Block> {
        public int height;
        public int i, j;
        public Block(int h, int i, int j) {
            this.height = h;
            this.i = i;
            this.j = j;
        }
        
        @Override
        public int compareTo(Block other) {
            if (this.height != other.height) {
                return this.height - other.height;
            } else {
                if (this.i != other.i) {
                    return this.i - other.i;
                } else {
                    return this.j - other.j;
                }
            }
        }
    }
    public int trapRainWater(int[][] heightMap) {
        int r = heightMap.length;
        int c = heightMap[0].length;

        PriorityQueue<Block> pq = new PriorityQueue<>();
        boolean[][] visited = new boolean[r][c];

        for (int i = 0; i < r; i++) {
            pq.offer(new Block(heightMap[i][0], i, 0));
            pq.offer(new Block(heightMap[i][c - 1], i, c - 1));
            visited[i][0] = visited[i][c - 1] = true;
        }
        for (int j = 1; j < c - 1; j++) {
            pq.offer(new Block(heightMap[0][j], 0, j));
            pq.offer(new Block(heightMap[r - 1][j], r - 1, j));
            visited[0][j] = visited[r - 1][j] = true;
        }

        int water = 0,
            dir[] = {-1, 0, 1, 0, -1};
        while (!pq.isEmpty()) {
            Block block = pq.poll();
            for (int d = 0; d < 4; d++) {
                int ni = block.i + dir[d];
                int nj = block.j + dir[d + 1];
                if (ni >= 0 && ni < r && nj >= 0 && nj < c && !visited[ni][nj]) {
                    water += Math.max(0, block.height - heightMap[ni][nj]);
                    pq.offer(new Block(Math.max(heightMap[ni][nj], block.height), ni, nj));
                    visited[ni][nj] = true;
                }
            }
        }

        return water;
    }
}

File: 407-trapping-rain-water-ii/trapping-rain-water-ii.cpp
class Solution {
public:
    int trapRainWater(vector<vector<int>>& heightMap) {
        if (heightMap.empty() || heightMap[0].empty()) return 0;
        int m = heightMap.size(), n = heightMap[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        priority_queue<pair<int, pair<int, int>>, vector<pair<int, pair<int, int>>>, greater<>> minHeap;
        for (int i = 0; i < m; ++i) {
            minHeap.push({heightMap[i][0], {i, 0}});
            minHeap.push({heightMap[i][n - 1], {i, n - 1}});
            visited[i][0] = visited[i][n - 1] = true;
        }
        for (int j = 1; j < n - 1; ++j) {
            minHeap.push({heightMap[0][j], {0, j}});
            minHeap.push({heightMap[m - 1][j], {m - 1, j}});
            visited[0][j] = visited[m - 1][j] = true;
        }
        
        int waterTrapped = 0;
        vector<int> directions = {0, 1, 0, -1, 0};
        while (!minHeap.empty()) {
            auto [height, cell] = minHeap.top();
            minHeap.pop();
            int x = cell.first, y = cell.second;
            for (int k = 0; k < 4; ++k) {
                int nx = x + directions[k], ny = y + directions[k + 1];
                if (nx >= 0 && nx < m && ny >= 0 && ny < n && !visited[nx][ny]) {
                    visited[nx][ny] = true;
                    waterTrapped += max(0, height - heightMap[nx][ny]);
                    minHeap.push({max(height, heightMap[nx][ny]), {nx, ny}});
                }
            }
        }

        return waterTrapped;
    }
};